File: photon-scala1/src/main/scala/photonscala/lib/LookAheadReader.scala
package photonscala.lib

import scala.collection.mutable

class LookAheadReaderBuffer[T](
  private val producer: () => T,
  private val buffer: mutable.ArrayDeque[T]
) {
  var bufferPointer = 0
  var readingFromBuffer = true

  def nextToken(): T = {
    if (readingFromBuffer && bufferPointer < buffer.length) {
      val token = buffer(bufferPointer)

      bufferPointer += 1

      token
    } else {
      readingFromBuffer = false

      val token = producer()

      buffer.append(token)

      token
    }
  }
}

class LookAheadReader[T](private val producer: () => T) {
  private val lookAheadBuffer = mutable.ArrayDeque[T]()

  def next(): T = {
    if (lookAheadBuffer.nonEmpty) {
      lookAheadBuffer.removeHead()
    } else {
      producer.apply()
    }
  }

  def lookAhead() = new LookAheadReaderBuffer(producer, lookAheadBuffer)
}

File: src/main/kotlin/photon/lib/LookAheadReader.kt
package photon.lib

data class LookAheadReaderBuffer<T>(
  private val producer: () -> T,
  private val buffer: ArrayDeque<T>
) {
  var bufferPointer = 0
  var readingFromBuffer = true

  fun nextToken(): T {
    if (readingFromBuffer && bufferPointer < buffer.size) {
      val token = buffer[bufferPointer]

      bufferPointer += 1

      return token
    } else {
      readingFromBuffer = false

      val token = producer()

      buffer.add(token)

      return token
    }
  }
}

class LookAheadReader<T>(
  private val producer: () -> T
) {
  private val lookAheadBuffer = ArrayDeque<T>()

  fun next(): T {
    if (lookAheadBuffer.isNotEmpty()) {
      return lookAheadBuffer.removeFirst()
    } else {
      return producer()
    }
  }

  fun lookAhead() = LookAheadReaderBuffer<T>(producer, lookAheadBuffer)
}